### ISSUE TITLE
cannot detect face

### ISSUE THREAD

**Issue by zhangxw87** [XiaXuehai/faceboxes#14]
I simply run inference.py with model weight/faceboxes.pt on images under picture/ folder, but cannot detect face on any of the images. what's the problem?

**Comment by XiaXuehai** [comment]
I tested well. what did `inference.py` print out.

**Comment by zhangxw87** [comment]
> I tested well. what did `inference.py` print out.

run python inference.py outputs
`There is no face in the image`

I did nothing except changing 
picture = '12.jpg'
to
picture = 'img_91.jpg'

**Comment by XiaXuehai** [comment]
![image](https://user-images.githubusercontent.com/11585321/47947507-9a153f80-df58-11e8-9938-a1142d348b7b.png)
did you use the master code?

**Comment by zhangxw87** [comment]
Yes. I used the master branch.

**Comment by XiaXuehai** [comment]
It is impossible. I downloaded the repo and tested again on my windows and linux system with the version of 0.4.1. 
Can you download the code and run it again? or upload the code.

**Comment by zhangxw87** [comment]
`
from networks import FaceBox
from encoderl import DataEncoder

import torch
import numpy as np
from torch.autograd import Variable
import torch.nn.functional as F
import cv2
import time


def detect(im):
    h, w, _ = im.shape
    dim_diff = np.abs(h - w)
    pad1, pad2 = dim_diff // 2, dim_diff - dim_diff // 2
    pad = ((pad1, pad2), (0, 0), (0, 0)) if h <= w else ((0, 0), (pad1, pad2), (0, 0))
    input_img = np.pad(im, pad, 'constant', constant_values=128)
    input_img = cv2.resize(input_img,(1024,1024))/255

    im_tensor = torch.from_numpy(input_img.transpose((2,0,1))).float()

    loc, conf = net(im_tensor.unsqueeze(0))

    boxes, labels, probs = data_encoder.decode(loc.data.squeeze(0),
                                                F.softmax(conf.squeeze(0), dim=1))
    if probs[0] != 0:
        boxes = boxes.numpy()
        probs = probs.detach().numpy()
        if h <= w:
            boxes[:,1] = boxes[:,1]*w-pad1
            boxes[:,3] = boxes[:,3]*w-pad1
            boxes[:,0] = boxes[:,0]*w
            boxes[:,2] = boxes[:,2]*w
        else:
            boxes[:,1] = boxes[:,1]*h
            boxes[:,3] = boxes[:,3]*h
            boxes[:,0] = boxes[:,0]*h-pad1
            boxes[:,2] = boxes[:,2]*h-pad1

    return boxes, probs

def testIm(file):
    im = cv2.imread(file)
    if im is None:
        print("can not open image:", file)
        return
    h,w,_ = im.shape
    boxes, probs = detect(im)

    if probs[0] == 0:
        print('There is no face in the image')
        exit()
    for i, (box) in enumerate(boxes):
        print('i=', i, 'box=', box)
        x1 = int(box[0])
        x2 = int(box[2])
        y1 = int(box[1])
        y2 = int(box[3])
        cv2.rectangle(im,(x1,y1+4),(x2,y2),(0,255,0),2)
        cv2.putText(im, str(probs[i]), (x1,y1), cv2.FONT_HERSHEY_COMPLEX, 0.4, (0,255,0))
    cv2.imshow('photo', im)
    cv2.imwrite('picture/1111.jpg', im)
    cv2.waitKey(0)


if __name__ == '__main__':
    net = FaceBox()
    net.load_state_dict(torch.load('weight/faceboxes.pt', map_location=lambda storage, loc:storage), strict=False) 
    net.eval()
    data_encoder = DataEncoder()

    # given image path, predict and show
    root_path = "picture/"
    picture = 'img_36.jpg'
    testIm(root_path + picture)
`
I used the above inference.py. Also tried to print out the shape of boxes, labeks, probs after line
 boxes, labels, probs = data_encoder.decode(loc.data.squeeze(0),
                                                F.softmax(conf.squeeze(0), dim=1)) 
they are tensor([0])

**Comment by XiaXuehai** [comment]
how about the `decode` function in `encoderl.py`?

**Comment by zhangxw87** [comment]
> how about the `decode` function in `encoderl.py`?

'    def decode(self, loc, conf):
        '''
        將预测出的 loc/conf转换成真实的人脸框
        loc [21842,4]
        conf [21824,2]
        '''
        variances = [0.1, 0.2]
        cxcy = loc[:,:2] * variances[0] * self.default_boxes[:,2:] + self.default_boxes[:,:2]
        wh  = torch.exp(loc[:,2:] * variances[1]) * self.default_boxes[:,2:]
        boxes = torch.cat([cxcy-wh/2,cxcy+wh/2],1) #[21824,4]

        # conf[:,0] means no face
        # conf[:,1] means face
        # filter face by a value of 0.4
        #conf[:,0] = 0.4

        max_conf, labels = conf.max(1) #[21842,1]

        if labels.long().sum().item() is 0:
            # no face in image
            return torch.tensor([0]), torch.tensor([0]), torch.tensor([0])

        ids = labels.nonzero().squeeze(1)
        # print('ids', ids)
        # print('boxes', boxes.size(), boxes[ids])

        keep = self.nms(boxes[ids],max_conf[ids])

        return boxes[ids][keep], labels[ids][keep], max_conf[ids][keep]
'

**Comment by XiaXuehai** [comment]
Sorry, I have no idea.The only thing I can think is the code is different.
you can send your code to my Email, ,or try to run it on the other machine.

**Comment by zhangxw87** [comment]
@XiaXuehai 
I think I have figured out the problem. 
division by 255 in 
https://github.com/XiaXuehai/faceboxes/blob/2b709e4115cf71d24603040a2ac517dd3c79a109/inference.py#L18
should be moved to 
https://github.com/XiaXuehai/faceboxes/blob/2b709e4115cf71d24603040a2ac517dd3c79a109/inference.py#L20

i.e. 
    input_img = cv2.resize(input_img,(1024,1024))
    im_tensor = torch.from_numpy(input_img.transpose((2,0,1))).float().div(255)

At least, it solves my problem.

**Comment by XiaXuehai** [comment]
......what's the difference?   0.0

**Comment by zhangxw87** [comment]
@XiaXuehai 
input_img becomes all 0 after being divided by 255.

**Comment by XiaXuehai** [comment]
did you use python 2?
